### DISCUSSION TITLE
How do I know whether to use `receive_text` or `receive_bytes`?

### DISCUSSION THREAD

**Discussion by WSH032** [frankie567/httpx-ws#52] [Q&A] [answered]
Hi, this is a great library!

I've been using this library recently to implement a reverse WebSocket proxy. I establish WebSocket connections with both the client and the target server, acting as an intermediary to pass messages between the two.

However, I've encountered an issue:

**Before receiving messages, how can I determine whether to use `receive_text` or `receive_bytes`? If I use the wrong method, I get a `WebSocketInvalidTypeReceived` error.**

```python
server.send_text("Hello World!")
# then
client.receive_bytes()  # Boom !
```

---

Actually, I know that I can achieve what I want by doing the following:
```python
import wsproto
from httpx_ws import WebSocketInvalidTypeReceived, AsyncWebSocketSession


async def receive(self: AsyncWebSocketSession) -> str | bytes:
    event = await self.receive()
    if isinstance(event, wsproto.events.TextMessage):
        return event.data
    elif isinstance(event, wsproto.events.BytesMessage):
        return event.data
    else:
        raise WebSocketInvalidTypeReceived(event)
```
But this isn't part of the official API, and I'm not certain whether future updates might cause disruptions.

---

I'm a bit confused. Why does `receive_bytes` only accept `bytes`, while `receive_text` only accepts `str`?

Because I cannot know in advance what type of data will be sent to me from the upstream.

> *[starlette/discussions/2310](https://github.com/encode/starlette/discussions/2310)*

Is this due to specific design purposes?

---

If I don't care about the type of message received, is there an officially supported API that allows receiving messages of any type, something like `async def receive(self: AsyncWebSocketSession) -> str | bytes:` I mentioned above?

And then I can determine the received data type by using `isinstance(data, str)`.

**Comment by frankie567** [comment, marked as answer]
Well, *generally speaking*, you know whether your WebSocket sends you a string or bytes; it's part of how it's implemented. So `receive_bytes` and `receive_text` are convenient shortcuts.

Obviously, for a proxy, you don't know in advance. Thus, the way you show with plain `receive` and check the nature of the message is the way to go. It's part of the API, it's documented: https://frankie567.github.io/httpx-ws/reference/httpx_ws/#httpx_ws.AsyncWebSocketSession.receive

**Reply by WSH032** [reply]
Thanks for your quick response!
---

> It's part of the API, it's documented:

I'm just a bit concerned because it involves `wsproto`, but `wsproto` is a dependency of `httpx_ws`, so I am not sure `wsproto` is part of `httpx_ws` API.
I worry that I might be digging a bit too deep into the lower-level APIs.

Thank you for your reply. I think the question has been resolved. :smile:
